File: cipher2.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;

public class HashTo6Digits {

    // 定义一个固定的盐值
    private static final byte[] SALT = "random_salt_value".getBytes(StandardCharsets.UTF_8);

    public static void main(String[] args) {
        // 获取系统当前日期并转换为 8 位数字
        int original = getCurrentDateAsOriginal();

        // 将 8 位数字加密为 6 位数字
        int encrypted = encryptTo6Digits(original);

        System.out.println("原始数字: " + original);
        System.out.println("加密结果(6位数字): " + encrypted);
    }

    public static int encryptTo6Digits(int num) {
        // 检查数字是否为 8 位
        if (num < 10000000 || num > 99999999) {
            throw new IllegalArgumentException("输入必须为 8 位数字");
        }

        try {
            // 转为字符串并计算哈希值
            String numStr = String.valueOf(num);

            // 使用 PBKDF2WithHmacSHA256 计算哈希值
            PBEKeySpec spec = new PBEKeySpec(numStr.toCharArray(), SALT, 100_000, 256);
            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            byte[] hashBytes = factory.generateSecret(spec).getEncoded();

            // 取前 4 个字节，转换为整数
            int hashInt = ((hashBytes[0] & 0xFF) << 24) | ((hashBytes[1] & 0xFF) << 16)
                    | ((hashBytes[2] & 0xFF) << 8) | (hashBytes[3] & 0xFF);

            // 限制结果为 6 位数字
            return Math.abs(hashInt % 1_000_000);
        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
            throw new RuntimeException("加密错误", e);
        }
    }

    public static int getCurrentDateAsOriginal() {
        // 获取当前日期并格式化为 YYYYMMDD
        LocalDate currentDate = LocalDate.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyyMMdd");
        return Integer.parseInt(currentDate.format(formatter));
    }
}

File: cipher2.py
import hashlib
import time

"""
需求： 只需要把8位数字压缩成6位数字,不需要考虑要不要进行解密还原，
一下子思路就不一样了，既然不需要还原，那么只需要单纯的加密哈希运算就行了。
"""

# 定义一个固定的盐值，建议更换为随机生成的安全盐值
SALT = b"random_salt_value"

def encrypt_to_6_digits(num):
    """将 8 位数字加密为 6 位数字"""
    if num < 10000000 or num > 99999999:
        raise ValueError("输入必须为 8 位数字")
    
    # 转为字符串并计算哈希值
    num_str = str(num).encode('utf-8')
    # hash_bytes = hashlib.sha256(dynamic_salt + num_str).digest()  # 加动态盐计算哈希值
    hash_bytes = hashlib.pbkdf2_hmac(
        'sha256',              # 使用的哈希算法
        num_str,               # 原始数据
        SALT,                  # 盐
        100_000                # 迭代次数（越高越安全，但计算越慢）
    )
    
    # 取前 4 个字节，转换为整数
    hash_int = int.from_bytes(hash_bytes[:4], 'big')
    
    # 限制结果为 6 位数字
    return hash_int % 10**6

# 获取系统当前日期并转换为 8 位数字
def get_current_date_as_original():
    current_time = time.localtime()
    return int(time.strftime("%Y%m%d", current_time))


# 示例
original = get_current_date_as_original()  # 输入 8 位数字
encrypted = encrypt_to_6_digits(original)

print(f"原始数字: {original}")
print(f"加密结果(6位数字): {encrypted}")
